File: src/main/java/com/warren/bt/Selector.java
package com.warren.bt;

public class Selector extends Composite {
  @Override
  public Status tick() {
    if (!hasChildren()) {
      return Status.FAILURE;
    }

    while (currentChild < children.size()) {
      Node child = children.get(currentChild);
      Status status = child.tick();

      if (status == Status.SUCCESS) {
        reset();
        return Status.SUCCESS;
      }

      if (status == Status.RUNNING) {
        return Status.RUNNING;
      }

      currentChild++;
    }

    reset();
    return Status.FAILURE;
  }
}

File: src/main/kotlin/com/warren/bt/Selector.kt
package com.warren.bt

class Selector : Composite("Selector") {
  private var currentIndex = 0

  override fun tick(): Status {
    while (currentIndex < children.size) {
      val status = children[currentIndex].tick()
      when (status) {
        Status.SUCCESS -> {
          return Status.SUCCESS
        }
        Status.FAILURE -> currentIndex++
        Status.RUNNING -> {
          return Status.RUNNING
        }
      }
    }
    return Status.FAILURE
  }

  override fun onReset() {
    currentIndex = 0
  }
}
